File: src/sim/parallel.ts
/**
 * Worker-pool batch runner for node.
 *
 * The paired harness's unit of meaning is the *block* — one deal played from
 * every seating — and a block's games depend only on its index (`simulate`'s
 * `blocks` option). So the pool hands each worker a slice of block indices,
 * the worker runs the ordinary serial loop over them, and the parent stitches
 * the results back in block order. Same seeds, same seatings, same stats as
 * the serial run — a test asserts byte-identical `computeStats` output.
 *
 * Agents cross the thread boundary as `{name, opts}` specs rebuilt by
 * `makeAgent` in the worker, since closures do not survive structured clone.
 * The convenient side effect: anything measurable must be registered in
 * src/agents/index.ts.
 *
 * Node-only (worker_threads); the web UI keeps its own single simWorker.
 */
import os from 'node:os';
import { Worker } from 'node:worker_threads';
import type { SimOptions, SimResult } from './runner';
import { permutations } from './runner';

/** How to build an agent on the far side of a thread boundary. */
export interface AgentSpec {
  name: string;
  opts?: Record<string, unknown>;
}

export interface ParallelSimOptions extends Omit<SimOptions, 'onGame'> {
  /** Thread count; defaults to all cores but one. */
  workers?: number;
  /** Time this agent's `choose` calls in-worker (index into the spec list). */
  timeAgentIndex?: number;
  onProgress?: (done: number, total: number) => void;
}

export interface ParallelSimResult extends SimResult {
  /** Wall-clock spent inside the timed agent's `choose`, if requested. */
  timing?: { ms: number; decisions: number };
}

interface WorkerDone {
  kind: 'done';
  games: SimResult['games'];
  timing: { ms: number; decisions: number } | null;
}
type WorkerMessage = { kind: 'progress' } | WorkerDone;

export async function simulateParallel(
  specs: AgentSpec[],
  opts: ParallelSimOptions,
): Promise<ParallelSimResult> {
  const rotate = opts.rotateSeats !== false;
  const paired = rotate && opts.paired !== false;
  const blockSize = paired ? permutations(specs.length).length : 1;
  const blockCount = Math.ceil(opts.games / blockSize);
  const totalGames = blockCount * blockSize;
  const workers = Math.max(1, Math.min(opts.workers ?? os.availableParallelism() - 1, blockCount));

  // Contiguous slices keep each worker's setup cache warm across neighbours.
  const slices: number[][] = Array.from({ length: workers }, () => []);
  for (let b = 0; b < blockCount; b++) slices[Math.floor((b * workers) / blockCount)].push(b);

  // Only plain data crosses the thread boundary: callbacks (onProgress here,
  // onDecision from PlayOptions) do not survive structured clone.
  const simOpts: Omit<SimOptions, 'onGame' | 'blocks' | 'onDecision'> = {
    players: opts.players,
    games: opts.games,
    seed: opts.seed,
    setupIndex: opts.setupIndex,
    setupMode: opts.setupMode,
    maxDecisions: opts.maxDecisions,
    rotateSeats: opts.rotateSeats,
    paired: opts.paired,
  };
  const { onProgress, timeAgentIndex } = opts;
  let done = 0;

  const results = await Promise.all(
    slices.map(
      (blocks) =>
        new Promise<WorkerDone>((resolve, reject) => {
          const worker = new Worker(new URL('./workerEntry.ts', import.meta.url), {
            workerData: { specs, opts: simOpts, blocks, timeAgentIndex: timeAgentIndex ?? null },
            execArgv: ['--import', 'tsx'],
          });
          worker.on('message', (m: WorkerMessage) => {
            if (m.kind === 'progress') onProgress?.(++done, totalGames);
            else resolve(m);
          });
          worker.on('error', reject);
          worker.on('exit', (code) => {
            if (code !== 0) reject(new Error(`sim worker exited with code ${code}`));
          });
        }),
    ),
  );

  const games = results.flatMap((r) => r.games).sort((a, b) => a.block - b.block);
  const timing =
    timeAgentIndex === undefined
      ? undefined
      : results.reduce(
          (acc, r) => ({
            ms: acc.ms + (r.timing?.ms ?? 0),
            decisions: acc.decisions + (r.timing?.decisions ?? 0),
          }),
          { ms: 0, decisions: 0 },
        );

  return { games, blockSize, paired, timing };
}

File: rust/crates/arcs-sim/src/parallel.rs
//! Partitioned batch runner. Ported from `src/sim/parallel.ts`.
//!
//! The paired harness's unit of meaning is the *block* — one deal played from
//! every seating — and a block's games depend only on its index
//! ([`SimOptions::blocks`]). So the pool hands each thread a slice of block
//! indices, the thread runs the ordinary serial loop over them, and the parent
//! stitches the results back in block order. Same seeds, same seatings, same
//! stats as the serial run: `tests/parallel.rs` asserts the two produce
//! **byte-identical** statistics.
//!
//! Agents cross the thread boundary as [`AgentSpec`]s rebuilt on the far side,
//! for the same reason the TS pool does it: a bot's state is not shareable, and
//! rebuilding it forces everything measurable to exist in the registry.
//!
//! The guarantee has one precondition, and it is worth naming because it is
//! the kind of thing that decays silently: **an agent's choice must not depend
//! on wall-clock time**. `mcts2-play` carries a per-decision `time_ms` budget
//! and therefore searches deeper on an idle machine than on a busy one; run it
//! through here and the parallel result will not reproduce the serial one. It
//! is an interactive preset, not a measurement preset, which is why the
//! gauntlet does not use it.

use std::sync::atomic::{AtomicUsize, Ordering};

use rayon::prelude::*;

use crate::runner::{
    AgentSpec, SeatedGame, SimError, SimOptions, SimResult, build_table, simulate_with_agents,
};

/// Threads to use when the caller does not say: all cores but one, so a long
/// batch leaves the machine usable.
pub fn default_workers() -> usize {
    std::thread::available_parallelism()
        .map(|n| n.get().saturating_sub(1).max(1))
        .unwrap_or(1)
}

/// Run a batch across `workers` threads.
///
/// `on_block` is called after each block finishes, with `(done, total)` block
/// counts. It runs on whichever worker finished the block, so it must be cheap
/// and thread-safe — and, per FINDINGS, it must **not** be used to time an
/// agent: wall-clock inside a saturated pool measures the pool. The gauntlet
/// samples thinking time serially for exactly that reason.
pub fn simulate_parallel(
    specs: &[AgentSpec],
    opts: &SimOptions,
    workers: usize,
    on_block: Option<&(dyn Fn(usize, usize) + Sync)>,
) -> Result<SimResult, SimError> {
    let block_count = match &opts.blocks {
        Some(list) => list.len(),
        None => opts.block_count(),
    };
    if block_count == 0 {
        return Ok(SimResult {
            games: Vec::new(),
            block_size: opts.block_size(),
            paired: opts.rotate_seats && opts.paired,
        });
    }
    let all_blocks: Vec<usize> = match &opts.blocks {
        Some(list) => list.clone(),
        None => (0..block_count).collect(),
    };
    let workers = workers.clamp(1, block_count);

    // Contiguous slices, as in the TS pool: neighbouring blocks share nothing
    // but they do share a code path, and keeping a worker on a run of them
    // keeps its caches warm.
    let mut slices: Vec<Vec<usize>> = vec![Vec::new(); workers];
    for (i, &b) in all_blocks.iter().enumerate() {
        slices[i * workers / block_count].push(b);
    }

    let done = AtomicUsize::new(0);
    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(workers)
        .build()
        .expect("rayon pool");

    let parts: Vec<Result<Vec<SeatedGame>, SimError>> = pool.install(|| {
        slices
            .par_iter()
            .map(|blocks| {
                // One table per worker, built once and reused across its
                // blocks — the agents' scratch buffers and search arenas are
                // the allocation the batch would otherwise repeat per game.
                let mut agents = build_table(specs)?;
                let mut out = Vec::new();
                for &block in blocks {
                    // One block at a time, so progress is reported at block
                    // granularity and a failure names the block that failed.
                    let sub = SimOptions {
                        blocks: Some(vec![block]),
                        ..opts.clone()
                    };
                    out.extend(simulate_with_agents(&mut agents, &sub)?.games);
                    if let Some(cb) = on_block {
                        cb(done.fetch_add(1, Ordering::Relaxed) + 1, block_count);
                    }
                }
                Ok(out)
            })
            .collect()
    });

    let mut games: Vec<SeatedGame> = Vec::new();
    for part in parts {
        games.extend(part?);
    }
    // Stable sort by block: within a block the seatings were pushed in `k`
    // order by the serial loop, and a stable sort preserves it. That is what
    // makes the reassembled batch identical to a serial one game for game.
    games.sort_by_key(|g| g.block);

    Ok(SimResult {
        games,
        block_size: opts.block_size(),
        paired: opts.rotate_seats && opts.paired,
    })
}
